rmw_dds_common/src/graph_cache.cpp

```diff
@@ -0,0 +1,966 @@
+  auto delete_reader_gid = [&](rmw_dds_common::msg::NodeEntitiesInfo & info)
+    {
+      auto it = std::find_if(
+        info.reader_gid_seq.begin(),
+        info.reader_gid_seq.end(),
+        [&](const rmw_dds_common::msg::Gid & gid)
+        {
+          return gid == reader_gid_msg;
+        });
+      if (it != info.reader_gid_seq.end()) {
+        info.reader_gid_seq.erase(it);
+      }
+    };
+
+  return __modify_node_info(
+    participant_gid, node_name, node_namespace, delete_reader_gid, participants_);
+}
+
+static
+rmw_ret_t
+__get_count(
+  const GraphCache::EntityGidToInfo & entities,
+  std::string topic_name,
+  size_t * count)
+{
+  assert(count);
+
+  *count = std::count_if(
+    entities.begin(),
+    entities.end(),
+    [&topic_name](const GraphCache::EntityGidToInfo::value_type & elem)
+    {
+      return elem.second.topic_name == topic_name;
+    });
+  return RMW_RET_OK;
+}
+
+rmw_ret_t
+GraphCache::get_writer_count(
+  const std::string & topic_name,
+  size_t * count) const
+{
+  std::lock_guard<std::mutex> guard(mutex_);
+  if (!count) {
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  return __get_count(data_writers_, topic_name, count);
+}
+
+rmw_ret_t
+GraphCache::get_reader_count(
+  const std::string & topic_name,
+  size_t * count) const
+{
+  std::lock_guard<std::mutex> guard(mutex_);
+  if (!count) {
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  return __get_count(data_readers_, topic_name, count);
+}
+
+static
+std::tuple<std::string, std::string, bool>
+__find_name_and_namespace_from_entity_gid(
+  const GraphCache::ParticipantToNodesMap & participant_map,
+  rmw_gid_t participant_gid,
+  rmw_gid_t entity_gid,
+  bool is_reader)
+{
+  auto it = participant_map.find(participant_gid);
+  if (participant_map.end() == it) {
+    return {"", "", false};
+  }
+  for (const auto & node_info : it->second) {
+    auto & gid_seq = is_reader ? node_info.reader_gid_seq : node_info.writer_gid_seq;
+    auto it = std::find_if(
+      gid_seq.begin(),
+      gid_seq.end(),
+      [&](const rmw_dds_common::msg::Gid & gid) -> bool {
+        return 0u == std::memcmp(gid.data.data(), entity_gid.data, RMW_GID_STORAGE_SIZE);
+      }
+    );
+    if (gid_seq.end() != it) {
+      return {node_info.node_name, node_info.node_namespace, true};
+    }
+  }
+  return {"", "", false};
+}
+
+using DemangleFunctionT = GraphCache::DemangleFunctionT;
+
+static
+rmw_ret_t
+__get_entities_info_by_topic(
+  const GraphCache::EntityGidToInfo & entities,
+  const GraphCache::ParticipantToNodesMap & participant_map,
+  const std::string & topic_name,
+  DemangleFunctionT demangle_type,
+  bool is_reader,
+  rcutils_allocator_t * allocator,
+  rmw_topic_endpoint_info_array_t * endpoints_info)
+{
+  assert(allocator);
+  assert(endpoints_info);
+
+  if (0u == entities.size()) {
+    return RMW_RET_OK;
+  }
+
+  // TODO(ivanpauno): Do something better than overallocating and shrinking to size.
```

Comment:
Can you expand on this TODO, and why can we not take care of it now?

Reply:
I've taken care of it, by first using `std::count`. I think that's way better than overallocating, though the ideal thing would be to have a way to resize a `rmw_topic_endpoint_info_array_t`. Or better, a way of doing `push_back`.

I can follow up on those, but I don't think it should be part of this PR.